service/resource_router_test.go

```diff
@@ -0,0 +1,95 @@
+package service
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/kvm-operator/service/resource/configmapv3"
+	"github.com/giantswarm/kvm-operator/service/resource/servicev2"
+	"github.com/giantswarm/operatorkit/framework"
+)
+
+func Test_Service_newResourceRouter(t *testing.T) {
+	resourceV1 := []framework.Resource{
+		&servicev2.Resource{},
+	}
+	resourceV2 := []framework.Resource{
+		&servicev2.Resource{},
+		&configmapv3.Resource{},
+	}
+
+	versionedResources := map[string][]framework.Resource{
+		"1.0.0": resourceV1,
+		"2.0.0": resourceV2,
+		"":      resourceV1,
+	}
+
+	testCases := []struct {
+		customObject      v1alpha1.KVMConfig
+		expectedResources []framework.Resource
+		errorMatcher      func(err error) bool
+	}{
+		// Test 0, get resource
+		{
+			customObject: v1alpha1.KVMConfig{
+				Spec: v1alpha1.KVMConfigSpec{
+					VersionBundle: v1alpha1.KVMConfigSpecVersionBundle{
+						Version: "1.0.0",
+					},
+				},
+			},
+			expectedResources: resourceV1,
+			errorMatcher:      nil,
+		},
+		// Test 1, get correct resources on multiples resources version
+		{
+			customObject: v1alpha1.KVMConfig{
+				Spec: v1alpha1.KVMConfigSpec{
+					VersionBundle: v1alpha1.KVMConfigSpecVersionBundle{
+						Version: "2.0.0",
+					},
+				},
+			},
+			expectedResources: resourceV2,
+			errorMatcher:      nil,
+		},
+		// Test 2, get resources from empty Version Bundle
+		{
+			customObject: v1alpha1.KVMConfig{
+				Spec: v1alpha1.KVMConfigSpec{
+					VersionBundle: v1alpha1.KVMConfigSpecVersionBundle{
+						Version: "",
+					},
+				},
+			},
+			expectedResources: resourceV1,
+			errorMatcher:      nil,
+		},
+		// Test 3, Invalid version returns an error.
+		{
+			customObject: v1alpha1.KVMConfig{
+				Spec: v1alpha1.KVMConfigSpec{
+					VersionBundle: v1alpha1.KVMConfigSpecVersionBundle{
+						Version: "",
+					},
+				},
+			},
+			expectedResources: resourceV1,
```

Comment:
IMO it does not make sense to expect a resource when you also expect an error. This test should actually fail when both is expected. It is also that test 2 and three test the same but in 3 you expect an error. That does not feel right.

Reply:
done